sdktests/server_side_persistence_base.go

```diff
@@ -0,0 +1,112 @@
+package sdktests
+
+import (
+	"time"
+
+	"github.com/redis/go-redis/v9"
+	"github.com/stretchr/testify/require"
+
+	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
+	"github.com/launchdarkly/go-sdk-common/v3/ldvalue"
+	"github.com/launchdarkly/go-server-sdk-evaluation/v3/ldbuilders"
+	"github.com/launchdarkly/sdk-test-harness/v2/framework/ldtest"
+	"github.com/launchdarkly/sdk-test-harness/v2/servicedef"
+)
+
+func doServerSidePersistenceTests(t *ldtest.T) {
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     "localhost:6379",
+		Password: "", // no password set
+		DB:       0,  // use default DB
+	})
+
+	newServerSidePersistenceTests(t, &RedisPersistenceStore{redis: rdb}).Run(t)
+}
+
+type ServerSidePersistenceTests struct {
+	persistence  PersistenceStore
+	initialFlags map[string]string
+}
+
+func newServerSidePersistenceTests(t *ldtest.T, persistence PersistenceStore) *ServerSidePersistenceTests {
+	flagKeyBytes, err :=
+		ldbuilders.NewFlagBuilder("flag-key").Version(1).
+			On(true).Variations(ldvalue.String("off"), ldvalue.String("match"), ldvalue.String("fallthrough")).
+			OffVariation(0).
+			FallthroughVariation(2).
+			Build().MarshalJSON()
+	require.NoError(t, err)
+
+	initialFlags := map[string]string{"flag-key": string(flagKeyBytes)}
+
+	uncachedFlagKeyBytes, err :=
+		ldbuilders.NewFlagBuilder("uncached-flag-key").Version(1).
+			On(true).Variations(ldvalue.String("off"), ldvalue.String("match"), ldvalue.String("fallthrough")).
+			OffVariation(0).
+			FallthroughVariation(2).
+			Build().MarshalJSON()
+	require.NoError(t, err)
+
+	initialFlags["uncached-flag-key"] = string(uncachedFlagKeyBytes)
+
+	return &ServerSidePersistenceTests{
+		persistence:  persistence,
+		initialFlags: initialFlags,
+	}
+}
+
+func (s *ServerSidePersistenceTests) Run(t *ldtest.T) {
+	t.Run("uses default prefix", s.usesDefaultPrefix)
+	t.Run("uses custom prefix", s.usesCustomPrefix)
+
+	t.Run("daemon mode", s.doDaemonModeTests)
+}
+
+func (s *ServerSidePersistenceTests) usesDefaultPrefix(t *ldtest.T) {
+	require.NoError(t, s.persistence.Reset())
```

Comment:
I think adding _just one_ level of wrapping (perhaps in `(s *ServerSidePersistenceTests) Run(t *ldtest.T)`)? 

could be useful: ensuring test setup/teardown happens the same for every test. Specifically, that `Reset` is called.

Maybe not, I could probably go either way. Also good to call out that these cannot run in parallel.

Reply:
One way we could solve it without wrapping/nesting/defer:

Make the API be like:
```
persistence := s.persistence.Open()

persistence.WriteData("foo", "bar")
```

where Open just calls Reset() and then returns the object (and open is the only function on `s.persistence` or whatever.)

Reply:
I expect this interface is going to change more as I add the other database integrations, and do the read-write types. 

Mind if I leave this as is, and then in a separate task, you and I both circle back together and see what kind of interface we can shake out?

Reply:
Yup sounds good